codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +314,498 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+
+                    if (bytes > 0) {
+                        promise.addListener(new ReturnFlowControlWindowOnFailureListener(bytes));
+                    }
+
+                    if (promise.isCancelled()) {
+                        // Write was cancelled so skip it.
+                        continue;
+                    }
+
+                    // TODO(buchgr): Should we also the change the writability if END_STREAM is set?
+                    try {
+                        doWrite(msg, promise);
+                    } catch (Throwable t) {
+                        promise.tryFailure(t);
+                    }
+                }
+            } finally {
+                // ensure we always flush even if the write loop throws.
+                ctx.flush();
+            }
+        }
+
+        /**
+         * Receive a read message. This does not notify handlers unless a read is in progress on the
+         * channel. May be called from any thread.
+         */
+        void fireChildRead(final Object msg) {
+            assert eventLoop().inEventLoop();
+
+            if (closed) {
+                ReferenceCountUtil.release(msg);
+                return;
+            }
+            if (readInProgress) {
+                assert inboundBuffer.isEmpty();
+                // Check for null because inboundBuffer doesn't support null; we want to be consistent
+                // for what values are supported.
+                RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+                readInProgress = doRead0(checkNotNull(msg, "msg"), allocHandle);
+                if (!allocHandle.continueReading()) {
+                    fireChildReadComplete();
+                }
+            } else {
+                inboundBuffer.add(msg);
+            }
+        }
+
+        void fireChildReadComplete() {
+            assert eventLoop().inEventLoop();
+
+            if (readInProgress) {
+                readInProgress = false;
+                unsafe().recvBufAllocHandle().readComplete();
+                pipeline().fireChannelReadComplete();
+            }
+        }
+
+        void incrementOutboundFlowControlWindow(int bytes) {
+            if (bytes == 0) {
+                return;
+            }
+            OUTBOUND_FLOW_CONTROL_WINDOW_UPDATER.addAndGet(this, bytes);
+        }
+
+        /**
+         * Returns whether reads should continue. The only reason reads shouldn't continue is that the
+         * channel was just closed.
+         */
+        private boolean doRead0(Object msg, RecvByteBufAllocator.Handle allocHandle) {
+            if (msg == CLOSE_MESSAGE) {
+                allocHandle.readComplete();
+                pipeline().fireChannelReadComplete();
+                close();
+                return false;
+            }
+            if (msg instanceof Http2WindowUpdateFrame) {
+                Http2WindowUpdateFrame windowUpdate = (Http2WindowUpdateFrame) msg;
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)